### DISCUSSION TITLE
How to use _setattr_ in Pydantic dataclasses when validate_assignment = True?

### DISCUSSION THREAD

**Discussion by decision-dev** [pydantic/pydantic#4351] [Question]
Pydantic dataclasses appear not to call __setattr__ on assignment when their config has validate_assignment=True.  They do call __setattr__ on both init and assignment when validate_assignment=False.  BaseModel appears to only call __setattr__ on assignmnet, but not on __init__ whether validate_assignment is True or False.  

What is the intended behavior?

Is there a way to effectively include code in the __setattr__ in all cases (on init and assignment,  whether validate_assignment is True or False, for dataclasses as well as BaseModel)?

```
from dataclasses import dataclass
from pydantic.dataclasses import dataclass as pydataclass
from pydantic.fields import ModelField
from pydantic import validator, BaseModel

@dataclass
class Test0(object):
    x: int

    def __setattr__(self, name, value):
        print('Called Test0 setattr with', name, value)
        super().__setattr__(name, value)
        
class Config:
    validate_assignment = True

@pydataclass(config=Config)
class Test1(object):
    x: int
    
    def __setattr__(self, name, value):
        print('Called Test1 setattr with', name, value)
        super().__setattr__(name, value)
        
@pydataclass()
class Test2(object):
    x: int
    
    def __setattr__(self, name, value):
        print('Called Test2 setattr with', name, value)
        super().__setattr__(name, value)

class Test3(BaseModel):
    x: int
    
    class Config:
        validate_assignment = True

    def __setattr__(self, name, value):
        print('Called Test3 setattr with', name, value)
        super().__setattr__(name, value)
    
class Test4(BaseModel):
    x: int
    
    def __setattr__(self, name, value):
        print('Called Test4 setattr with', name, value)
        super().__setattr__(name, value)



if __name__ == "__main__":
    print('python dataclass')
    test = Test0(x='5')
    print('init:', test)
    test.x = '3'
    print('assign:', test, '\n')
    
    print('pydantic dataclass with validate_assignemnt=True')
    test = Test1(x='5')
    print('init:', test)
    test.x = '3'
    print('assign:', test, '\n')
    
    print('pydantic dataclass with validate_assignemnt=False')
    test = Test2(x='5')
    print('init:', test)
    test.x = '3'
    print('assign:', test, '\n')
    
    print('BaseModel with validate_assignemnt=True')
    test = Test3(x='5')
    print('init:', test)
    test.x = '3'
    print('assign:', test, '\n')
    
    print('BaseModel with validate_assignemnt=False')
    test = Test4(x='5')
    print('init:', test)
    test.x = '3'
    print('assign:', test, '\n')
    
    ```

Results from the above code:

python dataclass
Called Test0 setattr with x 5
init: Test0(x='5')
Called Test0 setattr with x 3
assign: Test0(x='3') 

pydantic dataclass with validate_assignemnt=True
init: Test1(x=5)
assign: Test1(x=3) 

pydantic dataclass with validate_assignemnt=False
Called Test2 setattr with x 5
init: Test2(x=5)
Called Test2 setattr with x 3
assign: Test2(x='3') 

BaseModel with validate_assignemnt=True
init: x=5
Called Test3 setattr with x 3
assign: x=3 

BaseModel with validate_assignemnt=False
init: x=5
Called Test4 setattr with x 3
assign: x='3'

**Comment by decision-dev** [comment]
I dug into the code a little.  It looks like I'm asking for the changes in line 2,3, 13,14 & 15 below.  
I'm not sure of the etiquette.  Can I just submit a PR?

```
def _dataclass_validate_assignment_setattr(self: 'Dataclass', name: str, value: Any) -> None:
    if hasattr(self, '__setattr_pre_parse__'):
        self.__setattr_pre_parse__(self, name, value)
    if self.__pydantic_initialised__:
        d = dict(self.__dict__)
        d.pop(name, None)
        known_field = self.__pydantic_model__.__fields__.get(name, None)
        if known_field:
            value, error_ = known_field.validate(value, d, loc=name, cls=self.__class__)
            if error_:
                raise ValidationError([error_], self.__class__)

    if hasattr(self, '__setattr_post_parse__'):
        self.__setattr_post_parse__(self, name, value)
    else:
        object.__setattr__(self, name, value)

**Reply by samuelcolvin** [reply]
I don't really understand what you're asking for. What end user behaviour do you need?

Feel free to create a PR of it helps explain what you're suggesting.

**Reply by decision-dev** [reply]
I am using pydantic dataclasses.  In general, I still want to have __setattr__ available in those classes even when validate_assignment = True so that I can act on the instance during an assignment.  I believe that pydantic validation acts on the class (and the assigned values), but cannot act on the instance (please let me know if that is an incorrect understanding).  On assignment, after validation, I would like to use existing attribute values in the instance and further modify behavior based on the value.

In my current, specific use case, I have attributes that are lists of other dataclasses.  I would like to make assignment of a value to the list instead act as append in specific circumstances.  This way my "parser" can simply assign data to fields in the dataclass and the dataclass itself knows when to append to a list instead of overwriting the list.

Did that clarify it?  Or did I just make it all even less understandable? 
Is there a way to do this in pydantic dataclasses as is?

**Reply by decision-dev** [reply]
I'm trying to create a structure parallel to __post_init_post_parse__ and  __post_init__, but for __setattr__.  Does that help clarify?
